src/databroker_elasticsearch/elasticdocument.py

```diff
@@ -0,0 +1,76 @@
+#!/usr/bin/env python3
+
+"""\
+Class for translating dictionary document to Elasticsearch entry.
+"""
+
+from databroker_elasticsearch.converters import getconverter
+
+
+class ElasticDocument:
+    """Transform dictionary document according to the `docmap`.
+
+    Convert selected items in the input dictionary and return
+    them in a new output dictionary under the same or different
+    keys.  Skip items that have a value of ``None`` or which
+    convert to ``None``.
+
+    Parameters
+    ----------
+    docmap : list of tuples or list of lists
+        The specification of dictionary conversion provided as
+        a list of tuples ``(keyin, keyout, converter)``.
+        Each tuple sets ``docout[keyout] = converter(docin[keyin])``
+        with missing keys in docin silently ignored.
+        The tuples may be abbreviated to one or two entries which
+        will imply ``keyout = keyin`` and ``converter = noconversion``.
+        The `converter` can be either a callable or a string name
+        registered with the `getconverter` function.
+
+    Attributes
+    ----------
+    docmap : list of tuples
+        The finalized specification for dictionary conversion.
+        This equals the `docmap` argument of class initialization
+        except that abbreviated `docmap` entries are expanded and
+        string-type converters replaced with corresponding function.
+    """
+
+    def __init__(self, docmap):
```

Comment:
Should this take in the converter dictionary so users can provide their own?

Reply:
I don't see a need to support that at this time.
As it is user can add (or replace) converters with `register_converter` decorator.

Reply:
I would rather go with a "you pass us a dictionary with converters and we provide a default" than a register function. This allows users to inspect the registry for what they already have, reduces the amount of code (since we no longer need register commands) and allows users to use familiar dict syntax (eg `pop`, `update`, etc.).

Reply:
I suggest to give a try with the current API.  The enhancements can be done later if needed.
Note the `docmap` argument for `ElasticDocument.__init__` allows for any callable converter - not just those that are registered.  This also works in yaml docmap list which can have any converter function that dumps and loads with yaml:

[This loads with `load_callback` and `load_elasticindex`]
```yaml
# Example configuration for ElasticIndex and ElasticCallback factory
databroker-elasticsearch:
  host: localhost
  index: iss-db
  docmap:
    # brokername  [esname[=brokername]]  [converter[=noconversion]]
    # the mapping must produce ES name "_id"

    # registered converters are resolved with getconverter(name)
    - [uid, _id, str]
    - [time, date, toisoformat]

    # custom converter functions can be also used
    - [foo, bar, !!python/name:myconverters.footobar '']

```

Reply:
hmm ok